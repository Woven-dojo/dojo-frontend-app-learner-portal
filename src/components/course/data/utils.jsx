import {
  COURSE_AVAILABILITY_MAP,
  COURSE_MODES_MAP,
  COURSE_PACING_MAP,
  LICENSE_SUBSIDY_TYPE,
  OFFER_SUBSIDY_TYPE,
} from './constants';

export function hasCourseStarted(start) {
  const today = new Date();
  const startDate = new Date(start);
  return startDate && today >= startDate;
}

export function findUserEnrollmentForCourseRun({ userEnrollments, key }) {
  return userEnrollments.find(
    ({
      isEnrollmentActive,
      isRevoked,
      courseRunId,
    }) => (isEnrollmentActive && !isRevoked && courseRunId === key),
  );
}

export function isUserEntitledForCourse({ userEntitlements, courseUuid }) {
  return userEntitlements.some(({ courseUuid: uuid }) => uuid === courseUuid);
}

export function weeksRemainingUntilEnd(courseRun) {
  const today = new Date();
  const end = new Date(courseRun.end);
  const secondsDifference = Math.abs(end - today) / 1000;
  const days = Math.floor(secondsDifference / 86400);
  return Math.floor(days / 7);
}

export function hasTimeToComplete(courseRun) {
  return courseRun.weeksToComplete <= weeksRemainingUntilEnd(courseRun);
}

export function isArchived(courseRun) {
  if (courseRun.availability) {
    return courseRun.availability === COURSE_AVAILABILITY_MAP.ARCHIVED;
  }
  return false;
}

export function isCourseSelfPaced(pacingType) {
  return pacingType === COURSE_PACING_MAP.SELF_PACED;
}

export function isCourseInstructorPaced(pacingType) {
  return pacingType === COURSE_PACING_MAP.INSTRUCTOR_PACED;
}

export function createCourseInfoUrl({ baseUrl, courseKey }) {
  return `${baseUrl}/courses/${courseKey}/info`;
}

export const numberWithPrecision = (number, precision = 2) => number.toFixed(precision);

// See https://openedx.atlassian.net/wiki/spaces/WS/pages/1045200922/Enroll+button+and+Course+Run+Selector+Logic
// for more detailed documentation on course run selection and the enroll button.
export function getActiveCourseRun(course) {
  return course.courseRuns.find(courseRun => courseRun.uuid === course.advertisedCourseRunUuid);
}

export function getAvailableCourseRuns(course) {
  return course.courseRuns.filter(
    courseRun => courseRun.isMarketable && courseRun.isEnrollable && !isArchived(courseRun),
  );
}

export function findOfferForCourse(offers, catalogList = []) {
  const offerIndex = offers.findIndex((offer) => catalogList?.includes(offer.catalog));
  if (offerIndex !== -1) {
    return offers[offerIndex];
  }
  return null;
}

const getBestCourseMode = (courseModes) => {
  const {
    VERIFIED, PROFESSIONAL, NO_ID_PROFESSIONAL, AUDIT, HONOR,
  } = COURSE_MODES_MAP;
  /** Returns the 'highest' course mode available.
    *  Modes are ranked ['verified', 'professional', 'no-id-professional', 'audit', 'honor'] */
  if (courseModes.includes(VERIFIED)) {
    return VERIFIED;
  }
  if (courseModes.includes(PROFESSIONAL)) {
    return PROFESSIONAL;
  }
  if (courseModes.includes(NO_ID_PROFESSIONAL)) {
    return NO_ID_PROFESSIONAL;
  }
  if (courseModes.includes(AUDIT)) {
    return AUDIT;
  }
  if (courseModes.includes(HONOR)) {
    return HONOR;
  }
  return null;
};

export function findHighestLevelSeatSku(seats) {
  /** Returns the first seat found from the preferred course mode */
  if (!seats || seats.length <= 0) {
    return null;
  }
  const courseModes = seats.map(seat => seat.type);
  const courseMode = getBestCourseMode(courseModes);
  return seats.find(seat => seat.type === courseMode)?.sku;
}

export function shouldUpgradeUserEnrollment({
  userEnrollment,
  subscriptionLicense,
  enrollmentUrl,
}) {
  const isAuditEnrollment = userEnrollment?.mode === COURSE_MODES_MAP.AUDIT;
  if (isAuditEnrollment && subscriptionLicense && enrollmentUrl) {
    return true;
  }
  return false;
}

export function hasLicenseSubsidy(subsidy) {
  return subsidy?.subsidyType === LICENSE_SUBSIDY_TYPE;
}

export function hasOfferSubsidy(subsidy) {
  return subsidy?.subsidyType === OFFER_SUBSIDY_TYPE;
}

// Truncate a string to less than the maxLength characters without cutting the last word and append suffix at the end
export function shortenString(str, maxLength, suffix, separator = ' ') {
  if (str.length <= maxLength) { return str; }
  return `${str.substr(0, str.lastIndexOf(separator, maxLength))}${suffix}`;
}
