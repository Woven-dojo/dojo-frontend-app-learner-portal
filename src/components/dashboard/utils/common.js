export const setDashIfEmpty = (item, key, callback) => {
  if (!(key in item) || !item[key]) {
    return '-';
  }
  return callback(item[key]);
};

export const isElementInDOM = (step) => {
  const nodes = document.querySelectorAll(step.selector);
  return nodes.length > 0;
};
