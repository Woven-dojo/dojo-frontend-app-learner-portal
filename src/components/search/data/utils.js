// eslint-disable-next-line import/prefer-default-export
export const sortItemsByLabelAsc = items => (
  items.sort((a, b) => a.label.localeCompare(b.label))
);
