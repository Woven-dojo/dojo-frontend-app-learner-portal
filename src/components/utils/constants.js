export const QUERY_PARAMS = {
  q: 'search',
  s: 'sort',
  fDiff: 'difficultyLevels',
  fLan: 'languages',
  fLern: 'learningPaths',
  fDel: 'deliveryMethods',
  pAct: 'paginCurrentPage',
  pPer: 'paginPerPage',
};

export const SORT_OPTIONS_NAME = {
  RECOMENDED: 'recomended',
  ALPHABETICALLY: 'alphabetically',
  DIFFICULTY_ASC: 'difficultyASC',
  DIFFICULTY_DESC: 'difficultyDESC',
};

export const queryInit = {
  search: '',
  sort: SORT_OPTIONS_NAME.RECOMENDED,
  difficultyLevels: [],
  languages: [],
  learningPaths: [],
  deliveryMethods: [],
  paginCurrentPage: 1,
  paginPerPage: 12,
};
