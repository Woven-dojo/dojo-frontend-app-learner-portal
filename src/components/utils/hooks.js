import { useState, useRef, useEffect } from 'react';
import { useHistory } from 'react-router-dom';

export const useIsFirstRender = () => {
  const isMountRef = useRef(true);
  useEffect(() => {
    isMountRef.current = false;
  }, []);
  return isMountRef.current;
};

export function useUrlParams(param) {
  const history = useHistory();
  const url = new URLSearchParams(history.location.search);

  const urlParam = url.get(param);
  const [value, setValue] = useState(urlParam !== null ? urlParam : '');

  function setProps(newVal) {
    const currentUrl = new URLSearchParams(history.location.search);
    if (newVal.length === 0) {
      currentUrl.delete(param);
    } else {
      currentUrl.set(param, newVal);
    }
    history.replace({ pathname: history.location.pathname, search: currentUrl.toString() });
    setValue(newVal);
  }

  return [value, setProps];
}
