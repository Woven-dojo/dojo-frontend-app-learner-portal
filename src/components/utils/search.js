import React from 'react';
import {
  CONTENT_TYPE_COURSE, CONTENT_TYPE_PATHWAY,
  COURSE_TITLE, NUM_RESULTS_COURSE, NUM_RESULTS_PATHWAY,
  PATHWAY_TITLE,
} from '../search/constants';
import SearchCourseCard from '../search/SearchCourseCard';
import SearchPathwayCard from '../pathway/SearchPathwayCard';

export const getContentTypeFromTitle = (title) => {
  switch (title) {
      case COURSE_TITLE:
        return CONTENT_TYPE_COURSE;
      case PATHWAY_TITLE:
        return CONTENT_TYPE_PATHWAY;
      default:
          return null;
  }
};

export const getHitComponentFromTitle = (title) => {
  switch (title) {
      case COURSE_TITLE:
        return SearchCourseCard;
      case PATHWAY_TITLE:
        return SearchPathwayCard;
      default:
        return null;
  }
};

export const getNoOfResultsFromTitle = (title) => {
  switch (title) {
    case COURSE_TITLE:
      return NUM_RESULTS_COURSE;
    case PATHWAY_TITLE:
      return NUM_RESULTS_PATHWAY;
    default:
      return 0;
  }
};

export const getSkeletonCardFromTitle = (title) => {
  switch (title) {
      case COURSE_TITLE:
          // eslint-disable-next-line react/jsx-filename-extension
        return <SearchCourseCard.Skeleton />;
      case PATHWAY_TITLE:
        return <SearchPathwayCard.Skeleton />;
      default:
        return null;
  }
};

export const getNoResultsMessage = (title) => {
  const lowerCaseTitle = title.toLowerCase();
  return {
    messageTitle: `No ${lowerCaseTitle} were found to match your search results.`,
    messageContent: `Check out some popular ${lowerCaseTitle} below.`,
  };
};

export const getSearchErrorMessage = (title) => {
  const lowerCaseTitle = title.toLowerCase();
  return {
    messageTitle: `An error occured while finding ${lowerCaseTitle} that match your search.`,
    messageContent: 'Please try again later.',
  };
};
