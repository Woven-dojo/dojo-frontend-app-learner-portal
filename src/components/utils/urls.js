import { getConfig } from '@edx/frontend-platform/config';

/**
 * Given either an `enterpriseSlug` or an `enterpriseCustomerInviteKey`, returns a
 * URL to the dojo-auth-mfe. This is enterprise specific and aware
 *
 * @param {string} enterpriseSlug Slug of an enterprise customer
 * @param {string} enterpriseCustomerInviteKey UUID of an EnterpriseCustomerInviteKey
 * @returns URL of the enterprise proxy login page in the LMS.
 */
// eslint-disable-next-line import/prefer-default-export
export const getLoginUrl = (enterpriseSlug, enterpriseCustomerInviteKey) => {
  const loginUrl = new URL(`${getConfig().LMS_BASE_URL}/iam/login/`);
  const queryParams = new URLSearchParams();

  queryParams.append('next', global.location);

  if (enterpriseSlug) {
    queryParams.append('enterprise_slug', enterpriseSlug);

    // append enterpriseSlug as a subdomain if it is not already there
    if (loginUrl.hostname.split('.')[0] !== enterpriseSlug) {
      loginUrl.hostname = `${enterpriseSlug}.${loginUrl.hostname}`;
    }
  }

  if (enterpriseCustomerInviteKey) {
    queryParams.append('enterprise_customer_invite_key', enterpriseCustomerInviteKey);
  }

  return loginUrl.href;
};
