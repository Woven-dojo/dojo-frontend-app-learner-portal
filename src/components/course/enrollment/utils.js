/* eslint-disable import/prefer-default-export */
import { enrollButtonTypes } from '../data/constants';

const {
  ENROLL_DISABLED,
  TO_COURSEWARE_PAGE,
  TO_DATASHARING_CONSENT,
  TO_ECOM_BASKET,
  TO_VOUCHER_REDEEM,
  VIEW_ON_DASHBOARD,
} = enrollButtonTypes;

/**
 * The main logic to determine enrollment type (scenario).
 */
export function determineEnrollmentType({
  subsidyData: {
    subscriptionLicense, userSubsidy, enrollmentUrl, courseHasOffer,
  },
  isUserEnrolled,
  isEnrollable,
  isCourseStarted,
}) {
  if (isUserEnrolled) {
    return isCourseStarted ? TO_COURSEWARE_PAGE : VIEW_ON_DASHBOARD;
  }
  if (!isEnrollable) { return ENROLL_DISABLED; }
  if (!enrollmentUrl) { return ENROLL_DISABLED; }
  if (subscriptionLicense && userSubsidy) { return TO_DATASHARING_CONSENT; }
  if (subscriptionLicense && !userSubsidy) { return TO_ECOM_BASKET; }
  if (!subscriptionLicense && !courseHasOffer) { return TO_ECOM_BASKET; }
  if (!subscriptionLicense && courseHasOffer) { return TO_VOUCHER_REDEEM; }
  return ENROLL_DISABLED;
}
