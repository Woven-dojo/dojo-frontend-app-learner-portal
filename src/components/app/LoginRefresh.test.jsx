import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, act } from '@testing-library/react';
import { AppContext } from '@edx/frontend-platform/react';

import renderer from 'react-test-renderer';
import LoginRefresh from './LoginRefresh';
import * as utils from '../../utils/common';

jest.mock('../../utils/common');

// eslint-disable-next-line react/prop-types
const LoginRefreshWithContext = ({ roles = [] }) => (
  <AppContext.Provider value={{
    authenticatedUser: {
      userId: 1,
      roles,
    },
  }}
  >
    <LoginRefresh>
      <div>Hello!</div>
    </LoginRefresh>
  </AppContext.Provider>
); /* eslint-enable react/prop-types */

describe('<LoginRefresh />', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should call loginRefresh if the user has no roles', async () => {
    await act(async () => render(
      <LoginRefreshWithContext />,
    ));

    expect(utils.loginRefresh).toHaveBeenCalledTimes(1);
  });

  it('should not call loginRefresh if the user has roles', async () => {
    await act(async () => render(
      <LoginRefreshWithContext roles={['role-1']} />,
    ));

    expect(utils.loginRefresh).not.toHaveBeenCalled();
  });

  it('should render the expected HTML', async () => {
    await renderer.act(async () => {
      const tree = await renderer.create(
        <LoginRefreshWithContext />,
      );

      expect(tree.toJSON()).toMatchSnapshot();
    });
  });
});
