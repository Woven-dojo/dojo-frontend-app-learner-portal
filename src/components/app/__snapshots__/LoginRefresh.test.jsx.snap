// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginRefresh /> should render the expected HTML 1`] = `
<div
  className="py-5 container-mw-lg container-fluid"
>
  <div
    className="loading-spinner d-flex justify-content-center align-items-center"
  >
    <div
      className="spinner-border text-primary"
      role="status"
    >
      <span
        className="sr-only"
      >
        loading user details
      </span>
    </div>
  </div>
</div>
`;
